### PR TITLE
Add new test exclusive_scan.pass.cpp for USM shared/device memory

### PR DESCRIPTION
**Pull request by SergeyKopienko**
Added new test exclusive_scan.pass.cpp for USM shared/device memory:
- run test oneapi::dpl::exclusive_scan on USM shared/device memory;
- run test std::exclusive_scan;
- check (compare) results.

### CODE REVIEW

File: test/parallel_api/numeric/numeric.ops/exclusive_scan.pass.cpp [hunk 1]
```diff
@@ -0,0 +1,102 @@
...
+
+#include "support/utils.h"
+
+#if TEST_DPCPP_BACKEND_PRESENT
+
+#include "support/sycl_alloc_utils.h"
+
+template <sycl::usm::alloc alloc_type>
+void
+test_with_usm(sycl::queue& q, const ::std::size_t __count)
```
Comment (MikeDvorskiy) [inline test/parallel_api/numeric/numeric.ops/exclusive_scan.pass.cpp:32]: There is non consistency in the file  - uglified identificators or non-uglified.
Actually, we follow "uglification" just for the implementation, not for the test.

File: test/parallel_api/numeric/numeric.ops/exclusive_scan.pass.cpp [hunk 2]
```diff
@@ -0,0 +1,102 @@
...
+{
+    auto prepare_data = [__count](std::vector<::std::size_t>& idx, std::vector<int>& val)
+    {
+        for (int i = 0; i < __count; i++)
+        {
+            idx[i] = i + 1;
+            val[i] = 0;
```
Comment (MikeDvorskiy) [inline test/parallel_api/numeric/numeric.ops/exclusive_scan.pass.cpp:39]: I think we should explain why we fill an output sequence by zero.

File: test/parallel_api/numeric/numeric.ops/exclusive_scan.pass.cpp (hunk 1 above)
Comment (SergeyKopienko) [reply test/parallel_api/numeric/numeric.ops/exclusive_scan.pass.cpp]: Fixed

File: test/parallel_api/numeric/numeric.ops/exclusive_scan.pass.cpp [hunk 3]
```diff
@@ -0,0 +1,102 @@
...
+        }
+    };
+
+    // Prepare source data
+    std::vector<::std::size_t> h_idx(__count);
+    std::vector<int>           h_val(__count);
```
Comment (MikeDvorskiy) [inline test/parallel_api/numeric/numeric.ops/exclusive_scan.pass.cpp]: The are different types for the input and the output sequences.
Is it by design or a mistake? If by design what about integer overflow in the output sequence?

File: test/parallel_api/numeric/numeric.ops/exclusive_scan.pass.cpp [hunk 4]
```diff
@@ -0,0 +1,102 @@
...
+    // Copy source data to USM shared/device memory
+    TestUtils::usm_data_transfer<alloc_type, ::std::size_t> dt_helper_h_idx(q, ::std::begin(h_idx), ::std::end(h_idx));
+    auto d_idx = dt_helper_h_idx.get_data();
+
+    TestUtils::usm_data_transfer<alloc_type, int> dt_helper_h_val(q, ::std::begin(h_val), ::std::end(h_val));
+    auto d_val = dt_helper_h_val.get_data();
+
+    // Run dpl::exclusive_scan algorithm on USM shared-device memory
+    auto myPolicy = oneapi::dpl::execution::make_device_policy<
+        TestUtils::unique_kernel_name<class copy, (::std::size_t)alloc_type>>(q);
```
Comment (MikeDvorskiy) [inline test/parallel_api/numeric/numeric.ops/exclusive_scan.pass.cpp:57]: Type erasure casting.

Comment (SergeyKopienko) [reply test/parallel_api/numeric/numeric.ops/exclusive_scan.pass.cpp:57]: unique_kernel_name required second template param as std::size_t

File: test/parallel_api/numeric/numeric.ops/exclusive_scan.pass.cpp (hunk 3 above)
Comment (SergeyKopienko) [reply test/parallel_api/numeric/numeric.ops/exclusive_scan.pass.cpp]: It is mistake, fixed.

File: test/parallel_api/numeric/numeric.ops/exclusive_scan.pass.cpp (hunk 2 above)
Comment (SergeyKopienko) [reply test/parallel_api/numeric/numeric.ops/exclusive_scan.pass.cpp:39]: Fixed

File: test/parallel_api/numeric/numeric.ops/exclusive_scan.pass.cpp (hunk 4 above)
Comment (MikeDvorskiy) [reply test/parallel_api/numeric/numeric.ops/exclusive_scan.pass.cpp:57]: > unique_kernel_name required second template param as std::size_t

I understand it... But it is not a motivation for type erasure.

File: test/parallel_api/numeric/numeric.ops/exclusive_scan.pass.cpp [hunk 5]
```diff
@@ -0,0 +1,104 @@
...
+    oneapi::dpl::exclusive_scan(myPolicy, d_idx, d_idx + count, d_val, 0);
+
+    // Copy results from USM shared/device memory to host
+    std::vector<::std::size_t> h_sidx(count);
+    std::vector<int>           h_sval(count);
+    dt_helper_h_idx.retrieve_data(h_sidx.begin());
```
Comment (MikeDvorskiy) [inline test/parallel_api/numeric/numeric.ops/exclusive_scan.pass.cpp:63]: Why do we need the source data copying back?
Actually we should check just result.

File: test/parallel_api/numeric/numeric.ops/exclusive_scan.pass.cpp [hunk 6]
```diff
@@ -0,0 +1,104 @@
...
+    dt_helper_h_val.retrieve_data(h_sval.begin());
+
+    // Check results
+    std::vector<::std::size_t> h_sidx_expected(count);
```
Comment (MikeDvorskiy) [inline test/parallel_api/numeric/numeric.ops/exclusive_scan.pass.cpp:67]: I guess, here, we can use `h_idx  and  h_val`, initialized above.

**Review by MikeDvorskiy — CHANGES_REQUESTED** [review]

File: test/parallel_api/numeric/numeric.ops/exclusive_scan.pass.cpp (hunk 4 above)
Comment (SergeyKopienko) [reply test/parallel_api/numeric/numeric.ops/exclusive_scan.pass.cpp]: Fixed

File: test/parallel_api/numeric/numeric.ops/exclusive_scan.pass.cpp (hunk 5 above)
Comment (SergeyKopienko) [reply test/parallel_api/numeric/numeric.ops/exclusive_scan.pass.cpp]: Fixed

**Review by MikeDvorskiy — DISMISSED** [review]

File: test/parallel_api/numeric/numeric.ops/scan.pass.cpp [hunk 7]
```diff
@@ -27,69 +28,6 @@
 
 using namespace TestUtils;
 
-// We provide the no execution policy versions of the exclusive_scan and inclusive_scan due checking correctness result of the versions with execution policies.
```
Comment (MikeDvorskiy) [inline test/parallel_api/numeric/numeric.ops/scan.pass.cpp:30]: Availability the serial scanning algorithms in the C++ standard library depends on C++ version and version of the standard library. We should check these  versions and depending on that enable these serial versions of scan or don't - see //TODO on 31th line.

**Review by MikeDvorskiy — CHANGES_REQUESTED** [review]

File: test/parallel_api/numeric/numeric.ops/scan.pass.cpp (hunk 7 above)
Comment (SergeyKopienko) [reply test/parallel_api/numeric/numeric.ops/scan.pass.cpp:30]: OK, will fix

Comment (SergeyKopienko) [reply test/parallel_api/numeric/numeric.ops/scan.pass.cpp:30]: We already have self-implementer functions like inclusive_scan_serial, exclusive_scan_serial in scan.pass.cpp
I simply extract them to new header file.

File: test/parallel_api/numeric/numeric.ops/scan_serial_impl.h [hunk 8]
```diff
@@ -0,0 +1,84 @@
+// -*- C++ -*-
+//===-- scan_serial_impl.h ------------------------------------------------===//
```
Comment (MikeDvorskiy) [inline test/parallel_api/numeric/numeric.ops/scan_serial_impl.h:2]: //===----------------------------------------------------------------------===//

Comment (SergeyKopienko) [reply test/parallel_api/numeric/numeric.ops/scan_serial_impl.h:2]: Fixed. Will commit after completing all tests.

**Review by MikeDvorskiy — APPROVED** [review]

**Pull request merged**